awx/api/serializers.py

```diff
@@ -5356,24 +5357,67 @@ def validate(self, attrs):
 class InstanceLinkSerializer(BaseSerializer):
     class Meta:
         model = InstanceLink
-        fields = ('id', 'url', 'related', 'source', 'target', 'link_state')
+        fields = ('id', 'related', 'source', 'target', 'target_full_address', 'link_state')
 
     source = serializers.SlugRelatedField(slug_field="hostname", queryset=Instance.objects.all())
-    target = serializers.SlugRelatedField(slug_field="hostname", queryset=Instance.objects.all())
+
+    target = serializers.SerializerMethodField()
+    target_full_address = serializers.SerializerMethodField()
 
     def get_related(self, obj):
         res = super(InstanceLinkSerializer, self).get_related(obj)
         res['source_instance'] = self.reverse('api:instance_detail', kwargs={'pk': obj.source.id})
-        res['target_instance'] = self.reverse('api:instance_detail', kwargs={'pk': obj.target.id})
+        res['target_address'] = self.reverse('api:receptor_address_detail', kwargs={'pk': obj.target.id})
         return res
 
+    def get_target(self, obj):
+        return obj.target.instance.hostname
+
+    def get_target_full_address(self, obj):
+        return obj.target.get_full_address()
+
 
 class InstanceNodeSerializer(BaseSerializer):
     class Meta:
         model = Instance
         fields = ('id', 'hostname', 'node_type', 'node_state', 'enabled')
 
 
+class ReceptorAddressSerializer(BaseSerializer):
+    full_address = serializers.SerializerMethodField()
+
+    class Meta:
+        model = ReceptorAddress
+        fields = ('id', 'url', 'address', 'port', 'protocol', 'websocket_path', 'is_internal', 'instance', 'peers_from_control_nodes', 'full_address')
+        read_only_fields = ('full_address',)
+
+    def get_full_address(self, obj):
+        return obj.get_full_address()
+
+    def validate(self, attrs):
+        def get_field_from_model_or_attrs(fd):
+            return attrs.get(fd, self.instance and getattr(self.instance, fd) or None)
+
+        peers_from_control_nodes = get_field_from_model_or_attrs('peers_from_control_nodes')
+        instance = get_field_from_model_or_attrs('instance')
+
+        # only allow websocket_path to be set if protocol is ws
+        if attrs.get('protocol') != 'ws' and attrs.get('websocket_path'):
+            raise serializers.ValidationError(_("Can only set websocket path if protocol is ws."))
+
+        # an instance can only have one address with peers_from_control_nodes set to True
+        if peers_from_control_nodes:
+            for other_address in ReceptorAddress.objects.filter(instance=instance.id):
```

Comment:
I'd use a more compact queryset statement instead of this explicit loop.  Something like

```
if ReceptorAddress.objects.filter(instance=instance.id, peers_from_control_nodes=True).exists():
    raise ValidationError(...)
```

Reply:
What happens if the user is editing an existing address instead of creating a new one?  Won't either of these versions of the validation logic count the existing one?

Reply:
Probably correctable using an `and` clause on line 5409.  I forget what the logic is exactly for checking if a serializer is getting an existing object, but it probably has to do with seeing if the `id` is available and set on some part of the data.